xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp

```diff
@@ -290,40 +290,39 @@ int CAEEncoderFFmpeg::Encode(uint8_t *in, int in_size, uint8_t *out, int out_siz
 
     avcodec_fill_audio_frame(frame, channelNum, m_CodecCtx->sample_fmt, in, in_size, 0);
 
-    pkt->size = out_size;
-    pkt->data = out;
-
     /* encode it */
     err = avcodec_send_frame(m_CodecCtx, frame);
     if (err < 0)
       throw FFMpegException("Error sending a frame for encoding (error '{}')",
                             FFMpegErrorToString(err));
 
-    while (err >= 0)
+    err = avcodec_receive_packet(m_CodecCtx, pkt);
+    // err < 0 - we cannot cope with it
+    // err is EAGAIN or EOF - return to caller as well
+    if (err >= 0)
     {
-      err = avcodec_receive_packet(m_CodecCtx, pkt);
-      if (err == AVERROR(EAGAIN) || err == AVERROR_EOF)
+      if (pkt->size <= out_size)
       {
-        av_channel_layout_uninit(&frame->ch_layout);
-        av_frame_free(&frame);
-        av_packet_free(&pkt);
-        return (err == AVERROR(EAGAIN)) ? -1 : 0;
+        memset(out, 0, out_size);
```

Comment:
?
memset(out + pkt->size, 0, out_size - pkt->size);

Reply:
You cannot rely on the value of pkt->size, as you might never get a package. The stuff here works like this:
a) make a sane initial stage -> memset entire buffer
b) Copy the Buffer in, if we get one.

And
c) We break after first package received.

Your change would be perfectly right when we would expect more than one package, but then we also needed to check if it still fits.

Reply:
I guess I don't follow. The memcpy() which follows relies on value of pkt->size

apologies if I am making amateur mistake/assumption

Reply:
No, you have a point here. You don't touch the buffer if not packet is returned, therefore your save the memset. We play it safe and help us a bit with the debugger later on. 
Thanks for your comment.